### ISSUE TITLE
Non-null arguments

### ISSUE THREAD

**Issue by melinath** [melinath/django-graph-api#56]
We currently supports arguments ([see docs](http://facebook.github.io/graphql/October2016/#sec-Language.Arguments)) which are nullable by default. Marking an argument as non-null also makes it required.

### Requirements
For feature requirements, see the specifications for [NonNull types](http://facebook.github.io/graphql/October2016/#sec-Types.Non-Null) in section 3.1.8 (inputs only, not fields). Since the validation system is being done in #50, actual validation will be done later. For now, marking an argument as nullable e.g. `RelatedField(Episode, number=Int(null=False))`, should ensure the [introspection specifications](http://facebook.github.io/graphql/October2016/#sec-Type-Kinds.Non-Null) are met, specifically sections 4.2.2.8 and 4.2.2.9.

### Exclusions
- Validation for NonNull since that will be done in a separate issue after #50

**Comment by swimmingbean** [comment]
claiming this

**Comment by fugu13** [comment]
@swimmingbean let me know if you switch off of this, I paused on non-null fields and @ariannedee suggested I take a look at this one

**Comment by swimmingbean** [comment]
@fugu13 sure go for it

**Comment by fugu13** [comment]
Hmm, maybe come at this from a different direction: inside serialize

**Comment by fugu13** [comment]
The main downside is, right now serialize is very simple, and it would be strange to make NonNull's approach so different from List's, when the patterns for their serializations are basically identical

**Comment by ariannedee** [comment]
Hey @fugu13, have you made any progress on this? I was trying to figure out what you were saying in your comment and I think I get it your point about not wanting to change the tree structure.

I think I've managed to find a way around it. We can add a field to the GraphQL types (like on Scalar, and then InputValue in the future) that indicates it's required, and then when resolving the args, throw an error if need be. That handles the schema execution part.

And for introspection, we can fake the different tree structure. When we are resolving an arg `type`, return a NonNull type instead. The new NonNull type can take be instantiated with a type, like List is, and then return it as the `ofType`.

If you want to keep working on it, that's great. If not, let me know and I can do it. I'm not sure if this method is ideal and am open to other ones, but at least it gets around having to do meta programming.
